sound/soc/soc-pcm.c

```diff
@@ -2096,6 +2096,9 @@ int dpcm_be_dai_trigger(struct snd_soc_pcm_runtime *fe, int stream,
 		be = dpcm->be;
 		be_substream = snd_soc_dpcm_get_substream(be, stream);
 
+		if (!be_substream)
+			continue;
```

Comment:
that sounds really really bad, this means the trigger never worked, even with an older kernel.

Reply:
Agree on not segfaulting, but 
a) this is a year old so there's a very large validation issue here.
b) the root cause is not clear. a BE without substream is not something I can understand in an SOF context. I"d like more data to figure out with of the BEs in the topology is not properly configured or if it's by design.

@ranj063 FYI.

Reply:
@kv2019i if I look at all the be_substream uses, there are 2 cases where we test, and 4 where we might use an invalid pointer...

explicit test:
dpcm_be_dai_startup()
dpcm_apply_symmetry()
				      
Still problematic cases:
dpcm_be_dai_stop
https://github.com/thesofproject/linux/blob/3bdc36c2f22399de145ea4a03a8ee4be3fa8528f/sound/soc/soc-pcm.c#L1569

dpcm_be_dai_hw_params()
https://github.com/thesofproject/linux/blob/3bdc36c2f22399de145ea4a03a8ee4be3fa8528f/sound/soc/soc-pcm.c#L2015

dpcm_be_dai_trigger()
https://github.com/thesofproject/linux/blob/3bdc36c2f22399de145ea4a03a8ee4be3fa8528f/sound/soc/soc-pcm.c#L2121

dpcm_be_dai_hw_free()
https://github.com/thesofproject/linux/blob/3bdc36c2f22399de145ea4a03a8ee4be3fa8528f/sound/soc/soc-pcm.c#L1943

Reply:
Spending more time on this, I believe it's only possible to have a substream == NULL if the playback or capture substream count is zero

see https://github.com/thesofproject/linux/blob/3bdc36c2f22399de145ea4a03a8ee4be3fa8528f/sound/core/pcm.c#L650 
and 
https://github.com/thesofproject/linux/blob/3bdc36c2f22399de145ea4a03a8ee4be3fa8528f/sound/core/pcm.c#L667

So if you go up in the stack, this boils down to this:
https://github.com/thesofproject/linux/blob/3bdc36c2f22399de145ea4a03a8ee4be3fa8528f/sound/soc/soc-pcm.c#L2875

So my guess is that we end-up connecting FE and BE that are in different directions, and with no substream in a specific direction. This would point to a connection issue when creating the DPCM graph, and what we are seeing with the different if (!substream) are a result of a larger mistaken at a higher level.

my 2 cents.

Reply:
I strongly suspect a dailink configuration issue where we end-up capturing from a dailink which is only declared as a playback. The number of substreams really comes from the dpcm_playback and dpcm_capture fields at the dailink level.

https://github.com/thesofproject/linux/blob/3bdc36c2f22399de145ea4a03a8ee4be3fa8528f/sound/soc/soc-pcm.c#L2871

Reply:
@plbossart My commit description is at least wrong. So the "SSP1.IN" (see tplg diagram above) is a dummy BE that is actually never providing data, but I guess this was added to topology as some BE is needed. Actually data is coming from "demux", so there's a dummy FE-BE connection and actually data is going to FE<->FE.

Dmesg with additional logs like this.

opening echo ref PCM when no playback to speakers
[ 479.990379] EchoRef: ASoC: dpcm_be_connect FE EchoRef BE SSP1-Codec, besubstream 0000000000000000

doing the same when speakers (besubstream still NULL)
[ 501.655516] Speakers: ASoC: dpcm_be_connect FE Speakers BE SSP1-Codec, besubstream 000000009df83da6
[ 504.833614] EchoRef: ASoC: dpcm_be_connect FE EchoRef BE SSP1-Codec, besubstream 0000000000000000

Reply:
if you only start capture on the echoref (PCM6C) then no data is coming and it times out or just waits for data?
Is it so that one must start the playback first followed by the echoref capture to have this working as expected?

What is coming from the "ECHO REF 9"? It is strange that BUF9.0 is mixing two sources and SSP1.IN is not providing any data?

Is it so that DPCM needs a BE to be connected to a FE for a working stream? For the echoref capture you would only need a dummy BE. In DAPM domain this all should work (I think) since the capture would be connected to the playnack without issues: playback PCM is an active/enabled input, capture PCM is an active/enabled output after all and ll widgets in the path would be enabled.
But we are in DPCM domain, which is a bit different.

I guess you could have a 'running' FE without a BE stream, which would result  a stream which is idle, not moving anywhere (capture) or just dumping the data to /dev/null (playback), but clearly this happens.

Reply:
@ujfalusi @kv2019i So in that case, I believe we end-up with the PCM capture 6 and its matching FE that ends-up being connected with the BE Dailink SSP1-Codec. That link is only declared as playback, and when used for capture is really completely notional there is nothing happening on capture.

I asked @Vamshigopal to double-check if the oops goes away by declaring the SSP1 dailink with dpcm_capture = 1. In theory it should cause the substream to be allocated on capture as well, so there wouldn't be a NULL pointer used.

That said, the correct question is why we connect to a BE that's not supposed to handle the capture direction. Clearly we've built on a framework undocumented bug, possibly with the topology stuff just matching dailinks between topology and machine drivers without looking at the details, which means it can break at any time. Adding more checks might break other platforms though so it looks like a feature now.

I think the right set of fixes would be to declare this SSP1 dailink with capture, and to filter out for invalid/NULL substream across the board, not just in the trigger case not in the hw_params/free/stop that I listed above.

Reply:
> What is coming from the "ECHO REF 9"? It is strange that BUF9.0 is mixing two sources and SSP1.IN is not providing any data?

@ujfalusi it is a virtual widget meaning that it prevents the connection from SSP1.IN to BUF9.0.

Reply:
Ah yes @gongjun-song in hindsight @gongjun-song was right and I was wrong with PR #3753. The misunderstanding is that I didn't realize that the Chrome folks has already productized this topology with virtual connections, I mistakenly thought this was an addition for Brya only. My bad that we let this problem go through and ended-up firefighting after a kernel update.